Port from xml-rs to quick-xml

jcgruenhage opened the pull request:
Follow-up from #78. Things that happened since:
 - rebased this on the current changes on master
 - dropped the quick-xml impl into where the xml-rs one used to be, instead of offering quick-xml as a feature
 - fixed up the issue with pending collections
 - fixed a whole lot of clippy issues while I was at it

Thanks to @pr2502 for doing the heavy lifting here, most stuff was already in pretty good shape so I mostly just moved stuff around and did a few little touch-ups here and there.

jcgruenhage commented on rustfmt.toml:
```diff
@@ -1 +1 @@
-merge_imports = true
+imports_granularity="Crate"
```
rustfmt complained that merge_imports is deprecated and that I should use imports_granularity instead. It's in a separate commit together with all the clippy lints that have been fixed.

jcgruenhage commented on src/stream/binary_reader.rs:
```diff
@@ -418,7 +418,7 @@ mod tests {
         use crate::stream::Event::*;
 
-        let reader = File::open(&Path::new("./tests/data/binary.plist")).unwrap();
+        let reader = File::open(Path::new("./tests/data/binary.plist")).unwrap();
```
yep, there's a bunch of those, but those pre-existed. I just removed the referencing because clippy complained, but you're right of course, we don't need to create a Path here. I'll remove them.

jcgruenhage replied:
it's been done ^^

jcgruenhage commented on src/dictionary.rs:
```diff
@@ -14,6 +14,7 @@ use std::{
 use crate::Value;
 
 /// Represents a plist dictionary type.
+#[derive(Default, Clone, PartialEq)]
```
IndexMap does not provide an `Eq` implementation, so we can't derive `Eq` on it.

jcgruenhage commented on src/stream/xml_reader.rs:
```diff
@@ -1,221 +1,258 @@
+}
 
-    fn read_content(&mut self) -> Result<String, Error> {
+fn from_xml_error(err: XmlReaderError) -> ErrorKind {
```
This is indeed an error conversion, I've changed it to use `From` instead.

jcgruenhage commented on src/serde_tests.rs:
```diff
@@ -813,6 +813,15 @@ fn dictionary_deserialize_dictionary_in_struct() {
 }
 
+#[test]
+fn empty_array_and_dictionary() {
```
removed it

jcgruenhage commented on src/error.rs:
```diff
@@ -61,9 +61,10 @@ pub(crate) enum ErrorKind {
 pub(crate) enum FilePosition {
-    LineColumn(u64, u64),
+    // Never used anymore?
+    //LineColumn(u64, u64),
```
Simplified to a tuple struct for now

jcgruenhage commented on src/stream/xml_reader.rs:
```diff
@@ -1,221 +1,258 @@
-                        Some(ref open_name) if &name.local_name == open_name => (),
-                        Some(ref _open_name) => {
+                        Some(open_name) if name.local_name() == open_name.as_ref() => {}
+                        Some(_open_name) => {
```
Yep, makes sense, this looks a bit cleaner to me.

jcgruenhage commented:
Thanks for your feedback, @RickyDaMa. I've applied most of your suggestions, they do improve the code quite a bit. I'll be honest: I haven't completely read the code that I committed, just the bits I needed to touch to apply the feedback from over at #78 :sweat_smile:, as well as a bit required for #52 later on, although I haven't started implementation on that one yet, as I'd want this here and #85 to be merged first.

ebarnard commented:
@RickyDaMa Thanks for doing the review for me 😁! I'll try and take a look at it later today.

ebarnard commented on src/stream/binary_writer.rs:
```diff
@@ -527,7 +526,7 @@ impl<W: Write> BinaryWriter<W> {
                 } else {
                     let mut buf: [_; 9] = [0x87, 0, 0, 0, 0, 0, 0, 0, 0];
-                    buf[1..].copy_from_slice(&(v as u64).to_be_bytes());
+                    buf[1..].copy_from_slice(&v.to_be_bytes());
```
Please keep this, or change to `u64::to_be_bytes(v)`. I want it to be very clear what the type of `v` is.

jcgruenhage replied:
Part of the second commit, so this will be dropped out of the PR.

ebarnard commented on src/stream/binary_writer.rs:
```diff
@@ -713,24 +712,23 @@ mod tests {
 
         let streaming_parser = BinaryReader::new(Cursor::new(buf_inner));
 
-        let events: Vec<Result<_, _>> = streaming_parser.collect();
```
This collect into a `Vec` is intentional so that any errors that occur during reading are reported before any errors that occur during `Value::from_events`.

jcgruenhage replied:
Part of the second commit, so this will be dropped out of the PR.

ebarnard commented on src/serde_tests.rs:
```diff
@@ -848,34 +848,64 @@ fn dictionary_serialize_xml() {
-    let comparison = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>
-<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">
-<plist version=\"1.0\">
+    let comparison = r#"<?xml version="1.0" encoding="UTF-8"?>
```
I've had problems in the past with text editors replacing tabs with spaces. I'd rather keep the escaped version, even though it doesn't look very nice.

jcgruenhage replied:
converted it back, seems reasonable to me

ebarnard commented on src/stream/mod.rs:
```diff
@@ -96,9 +97,29 @@ impl XmlWriteOptions {
+            return self.with_indent(0, 0);
+        }
+
+        assert!(
```
I'm not really sure what to do in this case - I can't imagine anyone is using a non-repeating indent string - so maybe the panic is ok.

jcgruenhage replied:
Considering this is a public function, this is technically a breaking change for those who use a non-repeating indent string. I'd still argue that keeping it this way and having it panic in that case is still better than making it a breaking change for everyone.

I've added a doc string for `with_indent`, which I seem to have forgotten in my first round of refactoring, and added a deprecation warning for indent_string, to reduce the chance of people getting bitten by this.

ebarnard commented on src/stream/xml_reader.rs:
```diff
@@ -1,221 +1,262 @@
-                Ok(XmlEvent::StartElement { name, .. }) => {
+            if let Some(closed_name) = self.closed_element.take() {
+                // Check the corrent element is being closed
+                match self.element_stack.pop() {
```
[`Reader::check_end_names`](https://docs.rs/quick-xml/latest/quick_xml/reader/struct.Reader.html#method.check_end_names) looks like it might do this check for us.

jcgruenhage replied:
oh, and it's set to `true` by default, so we can skip this check completely. Good catch!

jcgruenhage replied:
I think that means we can get rid of everything related to `closed_element` and `element_stack`, which would reduce the code complexity quite a bit :)

jcgruenhage replied:
Yep, it does mean that, I've implemented it now.

ebarnard commented on src/stream/xml_reader.rs:
```diff
@@ -1,221 +1,262 @@
+                        b"true" => return Ok(Some(Event::Boolean(true))),
+                        b"false" => return Ok(Some(Event::Boolean(false))),
+
+                        b"array" | b"dict" => {
```
It looks like using [Reader::expand_empty_elements](https://docs.rs/quick-xml/latest/quick_xml/reader/struct.Reader.html#method.expand_empty_elements) can avoid us having to duplicate logic between `<array>...</array>` and `<array />` events.

jcgruenhage replied:
:+1:

ebarnard commented on src/stream/xml_reader.rs:
```diff
@@ -1,221 +1,262 @@
-                        );
-                    }
+                XmlEvent::Eof if self.element_stack.is_empty() => return Ok(None),
+                XmlEvent::Eof => return Err(self.with_pos(ErrorKind::UnclosedXmlElement)),
```
Again, I think we get this check for free if `Reader::check_end_names` is enabled via [`Error::UnexpectedEof`](https://docs.rs/quick-xml/latest/quick_xml/enum.Error.html).

ebarnard commented on src/stream/xml_reader.rs:
```diff
@@ -1,221 +1,262 @@
-    }
 }
 
 impl<R: Read> Iterator for XmlReader<R> {
```
I know I'm being a bit picky, but can this go above the impl block for `ReaderState`.

jcgruenhage replied:
Moved it up.

ebarnard commented on src/stream/xml_reader.rs:
```diff
@@ -1,221 +1,262 @@
+                    let pos = self.xml_reader_pos();
+                    let unespcaped = text
+                        .unescaped()
+                        .map_err(|err| ErrorKind::from(err).with_position(pos))?;
```
Does `.map_err(|e| self.with_pos(ErrorKind::from(e)))?;` work here?

jcgruenhage replied:
yep, it does!

ebarnard commented on src/stream/xml_writer.rs:
```diff
@@ -157,81 +144,125 @@ impl<W: Write> XmlWriter<W> {
     }
 
     fn write_boolean(&mut self, value: bool) -> Result<(), Error> {
+        self.handle_pending_collection()?;
```
Can this go in `write_value_event` so it's not duplicated across all the functions below?

jcgruenhage replied:
That needed a bit of restructuring, but yeah, good call.

ebarnard commented on src/stream/xml_writer.rs:
```diff
@@ -157,81 +144,125 @@ impl<W: Write> XmlWriter<W> {
         self.write_event(|this| {
-            match (this.stack.pop(), this.expecting_key) {
-                (Some(Element::Dictionary), true) => {
+            match (
```
This match is getting a bit hard to follow. I think you can just check beforehadn if there is a pending collection, and if so just write the empty tag and return before the existing match statement.

jcgruenhage replied:
split it up into two matches now

ebarnard reviewed (changes requested):
It's looking pretty good @jcgruenhage, just a few changes. I've only reviewed the first commit, please could you drop the clippy changes from this PR for the time being and put them in another once this is merged.

jcgruenhage commented on src/date.rs:
```diff
@@ -35,7 +35,7 @@ impl Date {
     }
 
     /// Converts the `Date` to an XML plist date string.
-    pub fn to_xml_format(&self) -> String {
```
It's going to be dropped anyway, because it's part of the second commit. But yes, stable API, good argument, even though it is just 16 bytes and it does implement `Copy`, which is why clippy does suggest changing that.e

jcgruenhage commented:
Thanks for the extensive review @ebarnard, I'll try to get the requested changes in today :)

jcgruenhage commented on src/stream/xml_reader.rs:
```diff
@@ -131,7 +131,7 @@ impl<R: Read> ReaderState<R> {
 
                 match closed_name.as_ref() {
                     b"array" | b"dict" => return Ok(Some(Event::EndCollection)),
-                    b"plist" | _ => {}
```
Part of the second commit, so this will be dropped out of the PR.

jcgruenhage commented on src/stream/xml_reader.rs:
```diff
@@ -216,7 +216,7 @@ impl<R: Read> ReaderState<R> {
 
                     match name.local_name() {
                         b"array" | b"dict" => return Ok(Some(Event::EndCollection)),
-                        b"plist" | _ => (),
```
Part of the second commit, so this will be dropped out of the PR.

jcgruenhage commented:
I've addressed all of the feedback except for the `Reader::check_end_names` thing so far, because that'll be a slightly bigger change. I hope I'll get to that later as well though.

jcgruenhage commented:
Okay, @ebarnard, your feedback should now be fully addressed, let me know if there is anything else you'd want changed here.

pr2502 commented:
hi @jcgruenhage thank you for finishing this! i'm happy you decided to tag me as a co-author, but may i ask you to replace my old tag with `max <36980911+pr2502@users.noreply.github.com>`. thanks a lot ^^

jcgruenhage commented:
Sure thing!

ebarnard commented on src/stream/xml_reader.rs:
```diff
@@ -1,219 +1,205 @@
+                ErrorKind::UnexpectedEof
             }
+            XmlReaderError::Io(err) => {
+                let err = if let Some(code) = err.raw_os_error() {
```
Can this not just be `XmlReaderError::Io(err) => ErrorKind::Io(err)`?

jcgruenhage replied:
apparently it can be, yeah. Some of the code that I didn't touch myself yet ^^

ebarnard commented on src/stream/xml_writer.rs:
```diff
@@ -157,28 +145,65 @@ impl<W: Write> XmlWriter<W> {
+                Some(PendingCollection::Array) => {
+                    this.xml_writer
+                        .write_event(XmlEvent::Empty(BytesStart::borrowed_name(b"array")))?;
+                    this.expecting_key = this.stack.last() == Some(&Element::Dictionary);
```
I don't think this, or the copy below is needed. `pending_collection` will only be `Some` if this is called straight after `write_start_dictionary` or `write_start_array` in which case the value of `expecting_key` will have remained unchanged so the existing value is always valid.

ebarnard replied:
No need to change this now by the way.

jcgruenhage replied:
I'd need to look into what is happening here in more detail to understand why it is how it is, but with those statements removed the tests fail.

ebarnard reviewed (commented):
Two more small questions and then we're good to go. Thank you so much @jcgruenhage for taking time to work on this.

ebarnard commented on src/stream/xml_writer.rs:
```diff
@@ -30,8 +27,12 @@ pub struct XmlWriter<W: Write> {
+    Dictionary,
 }
 
 impl<W: Write> XmlWriter<W> {
```
If you are changing anything else, please can you correct my type of `fearure` below!

jcgruenhage replied:
I'll include it as a separate commit :)

jcgruenhage commented:
@ebarnard I think this is ready now. One question: What's going to be the next released version? Asking because of the deprecation warning on `XmlWriteOptions::indent_string`

ebarnard commented:
Think it should be 1.4

jcgruenhage commented:
Then the deprecation is already good as it is. With your approval on it now, do you want to merge it?

lucasfernog commented:
@ebarnard do you have an idea when this could be released?

ebarnard commented:
v1.4 has now been released.
